cudax/include/cuda/experimental/__async/sender/atomic_intrusive_queue.cuh

```diff
@@ -0,0 +1,84 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//                         Copyright (c) 2023 Maikel Nadolski
+//===----------------------------------------------------------------------===//
+
+#ifndef __CUDAX_ASYNC_DETAIL_ATOMIC_INTRUSIVE_QUEUE
+#define __CUDAX_ASYNC_DETAIL_ATOMIC_INTRUSIVE_QUEUE
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/atomic>
+
+#include <cuda/experimental/__async/sender/intrusive_queue.cuh>
+#include <cuda/experimental/__detail/config.cuh>
+
+#include <cuda/experimental/__async/sender/prologue.cuh>
+
+namespace cuda::experimental::__async
+{
+template <auto _NextPtr>
+class _CCCL_TYPE_VISIBILITY_DEFAULT __atomic_intrusive_queue;
+
+template <class _Tp, _Tp* _Tp::* _NextPtr>
+class alignas(64) __atomic_intrusive_queue<_NextPtr>
+{
+public:
+  using __node_pointer        = _Tp*;
+  using __atomic_node_pointer = _CUDA_VSTD::atomic<_Tp*>;
+
+  [[nodiscard]]
+  _CUDAX_API auto empty() const noexcept -> bool
+  {
+    return __head_.load(_CUDA_VSTD::memory_order_relaxed) == nullptr;
+  }
+
+  _CUDAX_API auto push(__node_pointer __node) noexcept -> bool
+  {
+    _CCCL_ASSERT(__node != nullptr, "Cannot push a null pointer to the queue");
+    __node_pointer __old_head = __head_.load(_CUDA_VSTD::memory_order_relaxed);
+    do
+    {
+      __node->*_NextPtr = __old_head;
+    } while (!__head_.compare_exchange_weak(__old_head, __node, _CUDA_VSTD::memory_order_acq_rel));
+
+    if (__old_head != nullptr)
+    {
+      return false;
+    }
+
+    __head_.notify_one();
+    return true;
+  }
+
+  _CUDAX_API auto pop_all(bool __wait = true) noexcept -> __intrusive_queue<_NextPtr>
```

Comment:
Will this deadlock if someone uses `pop_all` an a queue and discards it?

```cpp
{
  auto queue = ...;
  
  queue.pop_all();
}
```
That sounds like a dangerous operation if someone is not too familiar with the implementation, especially given that waiting till something is in the queue is not really related to popping elements.

I would much prefer if we would drop the function argument and have a second method with a better name for that operation

Reply:
i have separated `pop_all(wait)` into `wait_for_item()` and `pop_all()`. i have also made `pop_all()` `[[nodiscard]]`.